### PR TITLE
refactor(storage): compact block 

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

New implementation of compact blocks
1. `BlockCompactMutator` iterates through the segments and selects the segments that need to be block compacted, the principle is the same as `segment compact`, set `block_per_seg` as the threshold, select the segments whose `block_count >= block_per_seg and block_count  < 2*block_per_seg` as a Part. So we get a set of Part.
2. Select the `row_count >= min_rows_per_block or block_size >= max_bytes_per_block` as the condition of `perfect_block` (`N` for short).
`CompactSource` gets a set of segments, iterates through the blocks, and finds the `block_statis >= N and block_statis < 2N` as a CompactTask. 
In this way, a set of `CompactTask` is obtained.
3. `CompactTransform` gets a set of `CompactTask`, only merge but not split, and generate new blocks and segments.
4. `CompactSink` gathers all the segments and commits them to the meta server.

In this process, add orders to keep the original order of segments and blocks.

The flow of Pipeline is as follows:

```
+--------------+        +-----------------+
|CompactSource1|  --->  |CompactTransform1|  ------
+--------------+        +-----------------+        |
|              |        |                 |        |      +-----------+
|    ...       |  ...   |       ...       |  ...   | ---> |CompactSink|
|              |        |                 |        |      +-----------+
+--------------+        +-----------------+        |
|CompactSourceN|  --->  |CompactTransformN|  ------
+--------------+        +-----------------+
```

Closes #issue

### CODE REVIEW

**Comment by BohuTANG** [comment]
Great.
@zhyass  Could you add more information about this PR(improvement or fixup), It's very useful for the user and the reviewer.

**Review by zhang2014 — APPROVED** [review]

File: src/query/catalog/src/table.rs [hunk 1]
```diff
@@ -247,7 +247,7 @@ pub trait Table: Sync + Send {
         target: CompactTarget,
         limit: Option<usize>,
         pipeline: &mut Pipeline,
-    ) -> Result<Option<Box<dyn TableMutator>>> {
+    ) -> Result<bool> {
```
Comment (Xuanwo) [inline src/query/catalog/src/table.rs:250]: How about adding doc comments for `compact` to explain the meaning of this `bool`?

**Comment by BohuTANG** [comment]
@dantengsky is carefully testing this PR :)

**Review by sundy-li — APPROVED** [review]

**Comment by Xuanwo** [comment]
> @dantengsky is carefully testing this PR :)

It's better to mark this PR as a draft to prevent it merged unintended

**Pull request merged**